File: external_apps/mobile_ios/mobile_ios/Core/Network/APIService.swift
import Foundation



class APIService {
    static let shared = APIService()
    private let tokenManager = TokenManager.shared
    
    private init() {}
    
    // MARK: - Generic Request
    func request<T: Codable>(
        endpoint: APIEndpoint,
        body: Encodable? = nil
    ) async throws -> T {
        guard let url = URL(string: APIConfig.apiURL + endpoint.path) else {
            throw NetworkError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = endpoint.method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        // Add JWT token if available
        if let token = tokenManager.getAccessToken() {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        
        // Add body if present
        if let body = body {
            request.httpBody = try JSONEncoder().encode(body)
        }
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw NetworkError.noData
        }
        
        if httpResponse.statusCode == 401 {
            throw NetworkError.unauthorized
        }
        
        guard (200...299).contains(httpResponse.statusCode) else {
            if let errorResponse = try? JSONDecoder().decode(APIResponse<String>.self, from: data) {
                throw NetworkError.serverError(errorResponse.message ?? "Server error")
            }
            throw NetworkError.serverError("HTTP \(httpResponse.statusCode)")
        }
        
        do {
            let decoded = try JSONDecoder().decode(T.self, from: data)
            return decoded
        } catch {
            throw NetworkError.decodingError
        }
    }
    
    // MARK: - Auth
    func login(email: String, password: String) async throws -> LoginResponse {
        let body = LoginRequest(email: email, password: password)
        let wrapper: APIResponse<LoginResponse> = try await request(endpoint: .login, body: body)
        
        guard let response = wrapper.data else {
            throw NetworkError.noData
        }
        
        // Save tokens
        tokenManager.saveAccessToken(response.accessToken)
        tokenManager.saveRefreshToken(response.refreshToken)
        tokenManager.saveUserId(response.user.id)
        
        return response
    }
    
    func getCurrentUser() async throws -> User {
        let response: APIResponse<User> = try await request(endpoint: .getCurrentUser)
        guard let user = response.data else {
            throw NetworkError.noData
        }
        return user
    }
    
    // MARK: - Departments
    func getDepartments() async throws -> [Department] {
        let response: APIResponse<[Department]> = try await request(endpoint: .getDepartments)
        return response.data ?? []
    }
    
    func createDepartment(_ department: Department) async throws -> Department {
        let response: APIResponse<Department> = try await request(endpoint: .createDepartment, body: department)
        guard let dept = response.data else {
            throw NetworkError.noData
        }
        return dept
    }
    
    func updateDepartment(id: String, _ department: Department) async throws -> Department {
        let response: APIResponse<Department> = try await request(endpoint: .updateDepartment(id: id), body: department)
        guard let dept = response.data else {
            throw NetworkError.noData
        }
        return dept
    }
    
    func deleteDepartment(id: String) async throws {
        let _: APIResponse<String> = try await request(endpoint: .deleteDepartment(id: id))
    }
    
    // MARK: - Employees
    func getEmployees() async throws -> [Employee] {
        let response: APIResponse<[Employee]> = try await request(endpoint: .getEmployees)
        return response.data ?? []
    }
    
    func createEmployee(_ employee: Employee) async throws -> Employee {
        let response: APIResponse<Employee> = try await request(endpoint: .createEmployee, body: employee)
        guard let emp = response.data else {
            throw NetworkError.noData
        }
        return emp
    }
    
    func updateEmployee(id: String, _ employee: Employee) async throws -> Employee {
        let response: APIResponse<Employee> = try await request(endpoint: .updateEmployee(id: id), body: employee)
        guard let emp = response.data else {
            throw NetworkError.noData
        }
        return emp
    }
    
    func deleteEmployee(id: String) async throws {
        let _: APIResponse<String> = try await request(endpoint: .deleteEmployee(id: id))
    }
}

File: external_apps/mobile_android/app/src/main/java/com/example/mobile_android/data/network/ApiService.kt
package com.example.mobile_android.data.network

import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.POST
import retrofit2.http.GET

interface ApiService {
    @POST("auth/login")
    suspend fun login(@Body body: Map<String, String>): Response<LoginResponse>

    @GET("auth/me")
    suspend fun getCurrentUser(): Response<ApiResponse<UserDto>>

    @GET("employees")
    suspend fun getEmployees(): Response<ApiResponse<List<EmployeeDto>>>

    @GET("employees/{id}")
    suspend fun getEmployee(@retrofit2.http.Path("id") id: String): Response<ApiResponse<EmployeeDto>>

    @POST("employees")
    suspend fun createEmployee(@Body employee: EmployeeDto): Response<ApiResponse<EmployeeDto>>

    @retrofit2.http.PUT("employees/{id}")
    suspend fun updateEmployee(@retrofit2.http.Path("id") id: String, @Body employee: EmployeeDto): Response<ApiResponse<EmployeeDto>>

    @retrofit2.http.DELETE("employees/{id}")
    suspend fun deleteEmployee(@retrofit2.http.Path("id") id: String): Response<ApiResponse<Unit>>

    @GET("departments")
    suspend fun getDepartments(): ApiResponse<List<DepartmentDto>>

    @POST("departments")
    suspend fun createDepartment(@Body department: DepartmentDto): ApiResponse<DepartmentDto>

    @retrofit2.http.PUT("departments/{id}")
    suspend fun updateDepartment(@retrofit2.http.Path("id") id: String, @Body department: DepartmentDto): ApiResponse<DepartmentDto>

    @retrofit2.http.DELETE("departments/{id}")
    suspend fun deleteDepartment(@retrofit2.http.Path("id") id: String): ApiResponse<Unit>

    @retrofit2.http.Multipart
    @POST("upload")
    suspend fun uploadFile(
        @retrofit2.http.Part file: okhttp3.MultipartBody.Part
    ): Response<ApiResponse<String>>
}

data class ApiResponse<T>(
    val success: Boolean,
    val data: T?,
    val message: String?
)

data class LoginResponse(
    val success: Boolean,
    val data: LoginData?,
    val message: String?
)

data class LoginData(
    val user: UserDto,
    val accessToken: String,
    val refreshToken: String,
    val require2FA: Boolean?,
    val tempToken: String?
)

data class UserDto(
    val id: String,
    val email: String,
    val fullName: String,
    val role: String
)

data class EmployeeDto(
    val id: String?,
    val firstName: String,
    val lastName: String,
    val email: String?,
    val position: String,
    val departmentId: String?
)

data class DepartmentDto(
    val id: String?,
    val name: String,
    val description: String?,
    val managerId: String?
)
